cocotb/share/lib/vhpi/VhpiCbHdl.cpp

```diff
@@ -548,9 +549,13 @@ int VhpiSignalObjHdl::set_signal_value(long value)
     switch (m_value.format) {
         case vhpiEnumVecVal:
         case vhpiLogicVecVal: {
+            if ((value > INT_MAX) || (value < INT_MIN)) {
+                LOG_ERROR("Data loss detected");
+            }
             int i;
+            vhpiEnumT value_as_enum = static_cast<vhpiEnumT>(value);
             for (i=0; i<m_num_elems; i++)
-                m_value.value.enumvs[m_num_elems-i-1] = value&(1L<<i);
+                m_value.value.enumvs[m_num_elems-i-1] = value_as_enum & vhpiEnumT(1 << i);
```

Comment:
Ias this ever correct? The more I look at it, the less I think it was.

Reply:
What is `decltype(m_value.value.enumvs)` here?

Reply:
`vhpiEnumT*` aka `unit32_t*`. The issues I see is how the simulator interprets the values being fed to it. Can I assume the simulator does a 0/!0 check, and assumes you only ever mean bit? There are more than two logic values. I think this code sections also covers the case of vectors of user-defined enums that aren't guaranteed to be binary, so you end up feeding it garbage.

Reply:
You're right that this should at least be casting to `bool` before performing the assignment

Reply:
For this and the related comments in this section I copied the implementation from `VhpiLogicSignalObjHdl::set_signal_value(long)` because it makes sense for this type of operation on this type of handle. I read through #333 and I'm still not sure what the original intent of this code was.